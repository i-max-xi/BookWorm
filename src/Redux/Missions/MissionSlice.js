const FETCH_MISSIONS = "space-travelers/Missions/FETCH_MISSIONS";
const baseURL = "https://api.spacexdata.com/v3/missions";

// Reducer
const missionsReducer = (state = [], action) => {
  switch (action.type) {
    case FETCH_MISSIONS:
      return action.missions;
    default:
      return state;
  }
};

// Action
export const getMissions = (missions) => ({
  type: FETCH_MISSIONS,
  missions,
});

// Fetch API
export const fetchMissions = () => async (dispatch) => {
  const arrayOfMissions = await fetch(baseURL)
    .then((res) => res.json())
    .then((data) =>
      Object.entries(data).map((mission) => {
        const { description, id } = mission[1];
        const name = mission[1].mission_name;
        return {
          id,
          description,
          name,
        };
      })
    );

  dispatch(getMissions(arrayOfMissions));
};

export default missionsReducer;
