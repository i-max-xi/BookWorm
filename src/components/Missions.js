import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchMissions } from '../Redux/Missions/MissionSlice';

const Missions = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  const missions = useSelector((state) => state.missions);

  return (
    <div className="missionsPage">
      <table>
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
            {/* Added an empty div inside this last table header, otherwise we get
          an error in eslint */}
            <th>
              <div />
            </th>
          </tr>
        </thead>
        <tbody>
          {missions.map((mission) => (
            <tr key={mission.id}>
              <td key={mission.id}>{mission.name}</td>
              <td key={mission.id}>{mission.description}</td>
              <td key={mission.id}>
                <button type="button">
                  {mission.reserved ? 'Active Member' : 'Not A Member'}
                </button>
              </td>
              <td key={mission.id}>
                <button type="button">
                  {mission.reserved ? 'Leave Mission' : 'Join Mission'}
                </button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default Missions;
