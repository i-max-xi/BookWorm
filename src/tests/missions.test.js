import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from '../Redux/store';
import Missions from '../components/Missions';

it('renders correctly', () => {
  const mission = renderer.create(
    <Provider store={configureStore}>
      <Missions />
    </Provider>,
  );
  expect(mission).toMatchSnapshot();
});
