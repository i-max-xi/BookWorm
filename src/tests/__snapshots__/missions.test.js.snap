// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="missionsPage"
>
  <table>
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
          <div />
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
